### ISSUE TITLE
[N2439] Clang does NOT issue a diagnostic on ill-formed overloaded member functions

### ISSUE THREAD

**Issue by llvmbot** [llvm/llvm-project#41340]
|  |  |
| --- | --- |
| Bugzilla Link | [41995](https://llvm.org/bz41995) |
| Version | 8.0 |
| OS | All |
| Reporter | LLVM Bugzilla Contributor |
| CC | @dwblaikie,@DougGregor,@zygoloid |

## Extended Description 
# Overview:

According to [N2439](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2007/n2439.htm),
> Member function declarations with the same name and the same parameter-type-list, as well as member function template declarations with the same name, the same parameter-type-list, and the same template parameter lists, cannot be overloaded if any of them, but not all, have a ref-qualifier (8.3.5).

# Steps to Reproduce:
```cpp
class Base {
public:
  void f() const {}
};

class Derived : public Base {
public:
  using Base::f;
  void f() & {}
  void f() && {}
};


int main() {
  const Derived cd;
  cd.f();
  
  return 0;
}

```

# Actual Results: 
Clang compiles the above example without a diagnostic.

# Expected Results:
Clang issues a diagnostic.

# Build Date & Hardware:
```
clang version 8.0.0 (tags/RELEASE_800/final)
Target: x86_64-unknown-linux-gnu
Thread model: posix
```
# Additional Information:
- godbolt link to explore the example online.
  - https://godbolt.org/z/vEYRIU

- x86-64 gcc 9.1 issues a diagnostic on the above example.
```
<source>: In function 'int square(int)':
<source>:15:10: error: passing 'const Derived' as 'this' argument discards qualifiers [-fpermissive]
   15 |     cd.f();
      |          ^
<source>:9:7: note:   in call to 'int Derived::f() &'
    9 |   int f() & { return 2; }
      |       ^
Compiler returned: 1
```

**Comment by llvmbot** [comment]
@llvm/issue-subscribers-clang-frontend

Author: None (llvmbot)

<details>
|  |  |
| --- | --- |
| Bugzilla Link | [41995](https://llvm.org/bz41995) |
| Version | 8.0 |
| OS | All |
| Reporter | LLVM Bugzilla Contributor |
| CC | @dwblaikie,@DougGregor,@zygoloid |

## Extended Description 
# Overview:

According to [N2439](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2007/n2439.htm),
&gt; Member function declarations with the same name and the same parameter-type-list, as well as member function template declarations with the same name, the same parameter-type-list, and the same template parameter lists, cannot be overloaded if any of them, but not all, have a ref-qualifier (8.3.5).

# Steps to Reproduce:
```cpp
class Base {
public:
  void f() const {}
};

class Derived : public Base {
public:
  using Base::f;
  void f() &amp; {}
  void f() &amp;&amp; {}
};


int main() {
  const Derived cd;
  cd.f();
  
  return 0;
}

```

# Actual Results: 
Clang compiles the above example without a diagnostic.

# Expected Results:
Clang issues a diagnostic.

# Build Date &amp; Hardware:
```
clang version 8.0.0 (tags/RELEASE_800/final)
Target: x86_64-unknown-linux-gnu
Thread model: posix
```
# Additional Information:
- godbolt link to explore the example online.
  - https://godbolt.org/z/vEYRIU

- x86-64 gcc 9.1 issues a diagnostic on the above example.
```
&lt;source&gt;: In function 'int square(int)':
&lt;source&gt;:15:10: error: passing 'const Derived' as 'this' argument discards qualifiers [-fpermissive]
   15 |     cd.f();
      |          ^
&lt;source&gt;:9:7: note:   in call to 'int Derived::f() &amp;'
    9 |   int f() &amp; { return 2; }
      |       ^
Compiler returned: 1
```
</details>

**Comment by cor3ntin** [comment]
https://godbolt.org/z/cTo9cnd68
I think clang is correct per https://eel.is/c++draft/basic.scope.scope#3 - but i would not mind an extra pair of eyes as this wording change a lot recently
@shafik @AaronBallman

**Comment by shafik** [comment]
At first I want to say clang is correct and gcc is the only once to object: https://godbolt.org/z/YG9W9eb56

Sadly we can't get consistent behavior on the four implementations if we make the `Derived` member functions `const` in turn:

if `f() &` is const edg/MSVC reject: https://godbolt.org/z/fr96W9z86

if `f() &&` is const gcc/clang reject: https://godbolt.org/z/TeaG5KxE7

Someone is wrong here on some of these cases but not sure whom.

CC @hubert-reinterpretcast @zygoloid

**Comment by hubert-reinterpretcast** [comment]
The path to the wording that @cor3ntin points to is https://wg21.link/namespace.udecl#11.

Clang is right to accept the original case (and call `f() const` from the base class). GCC is wrong to ignore the candidate from the base class: The declarations do not correspond because the object parameter types, after removing top-level references, differ in constness.
GCC and Clang are right to accept the `f() const &` case (and call `f() const &` from the derived class). EDG and MSVC are wrong to reject because the function from the base class is not a candidate to begin with: The declarations _do_ correspond.
GCC and Clang are right to reject the `f() const &&` case. EDG and MSVC are wrong to accept because the declarations _do_ correspond, meaning that we only have the non-viable candidates from the derived class.
